packages/ra-ui-materialui/src/input/NumberInput.tsx

```diff
@@ -113,7 +118,10 @@ export const NumberInput = ({
         hasFocus.current = true;
     };
 
-    const handleBlur = () => {
+    const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
+        if (!event.currentTarget.value && inputRef.current) {
```

Comment:
I'm afraid `!event.currentTarget.value` will return `false` when the value is `0`, which will reset the value although `0` is a valid value for `NumberInput`.